File: scripts/create-admin.js
#!/usr/bin/env node
/**
 * Admin Bootstrap Script for Internship Tracker
 * 
 * Usage:
 *   node scripts/create-admin.js --gitlab-username=amruthjakku --name="Amruth Jakku" --email="amruth@example.com"
 * 
 * This script creates the first admin user in the database.
 */

const mongoose = require('mongoose');
const path = require('path');
const fs = require('fs');

// Load environment variables
function loadEnvFile() {
  const envPath = path.join(__dirname, '..', '.env.local');
  if (fs.existsSync(envPath)) {
    const envContent = fs.readFileSync(envPath, 'utf8');
    envContent.split('\n').forEach(line => {
      const [key, value] = line.split('=');
      if (key && value) {
        process.env[key] = value;
      }
    });
  }
}

// User schema (simplified for script)
const userSchema = new mongoose.Schema({
  gitlabUsername: { type: String, required: true, unique: true },
  gitlabId: { type: String, required: true, unique: true },
  name: { type: String, required: true },
  email: { type: String, required: true },
  role: { type: String, required: true, enum: ['admin', 'mentor', 'intern'] },
  assignedBy: { type: String, required: true },
  isActive: { type: Boolean, default: true },
  profileImage: { type: String, default: null },
  lastLoginAt: { type: Date, default: null },
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now }
});

const User = mongoose.model('User', userSchema);

async function createAdmin(gitlabUsername, name, email, gitlabId = null) {
  try {
    // Load environment variables
    loadEnvFile();
    
    // Connect to MongoDB
    const mongoUri = process.env.MONGODB_URI || 'mongodb://localhost:27017/internship_tracker';
    await mongoose.connect(mongoUri);
    
    console.log('✅ Connected to MongoDB');
    
    // Check if admin already exists
    const existingAdmin = await User.findOne({
      gitlabUsername: gitlabUsername.toLowerCase()
    });
    
    if (existingAdmin) {
      console.log(`❌ Admin with GitLab username '${gitlabUsername}' already exists!`);
      return false;
    }
    
    // Check if any admin exists
    const existingAdminsCount = await User.countDocuments({ role: 'admin' });
    
    // Create admin user
    const adminUser = new User({
      gitlabUsername: gitlabUsername.toLowerCase(),
      gitlabId: gitlabId || `admin_${gitlabUsername.toLowerCase()}`,
      name: name,
      email: email.toLowerCase(),
      role: 'admin',
      assignedBy: 'system_bootstrap',
      isActive: true,
      profileImage: null,
      lastLoginAt: null
    });
    
    // Save admin user
    const savedUser = await adminUser.save();
    
    console.log('✅ Successfully created admin user!');
    console.log(`   GitLab Username: ${gitlabUsername}`);
    console.log(`   Name: ${name}`);
    console.log(`   Email: ${email}`);
    console.log(`   Role: admin`);
    console.log(`   User ID: ${savedUser._id}`);
    
    if (existingAdminsCount === 0) {
      console.log('🎉 This is the first admin user in the system!');
    } else {
      console.log(`📊 Total admins in system: ${existingAdminsCount + 1}`);
    }
    
    return true;
    
  } catch (error) {
    console.error('❌ Error creating admin:', error.message);
    return false;
  } finally {
    await mongoose.disconnect();
  }
}

function parseArgs() {
  const args = process.argv.slice(2);
  const parsed = {};
  
  args.forEach(arg => {
    if (arg.startsWith('--')) {
      const [key, value] = arg.substring(2).split('=');
      parsed[key.replace(/-/g, '_')] = value;
    }
  });
  
  return parsed;
}

async function main() {
  const args = parseArgs();
  
  if (!args.gitlab_username || !args.name || !args.email) {
    console.log('❌ Missing required arguments!');
    console.log('');
    console.log('Usage:');
    console.log('  node scripts/create-admin.js --gitlab-username=USERNAME --name="FULL NAME" --email="EMAIL"');
    console.log('');
    console.log('Example:');
    console.log('  node scripts/create-admin.js --gitlab-username=amruthjakku --name="Amruth Jakku" --email="amruth@example.com"');
    process.exit(1);
  }
  
  console.log('🚀 Creating admin user for Internship Tracker...');
  console.log(`   GitLab Username: ${args.gitlab_username}`);
  console.log(`   Name: ${args.name}`);
  console.log(`   Email: ${args.email}`);
  console.log('');
  
  const success = await createAdmin(
    args.gitlab_username,
    args.name,
    args.email,
    args.gitlab_id
  );
  
  if (success) {
    console.log('');
    console.log('🎯 Next Steps:');
    console.log('1. Start your Next.js application: npm run dev');
    console.log('2. Login with your GitLab account');
    console.log('3. You\'ll be automatically recognized as an admin');
    console.log('4. Use the admin dashboard to add mentors and colleges');
    process.exit(0);
  } else {
    console.log('');
    console.log('❌ Failed to create admin user. Please check the error above.');
    process.exit(1);
  }
}

main().catch(console.error);

File: scripts/create_admin.py
#!/usr/bin/env python3
"""
Admin Bootstrap Script for Internship Tracker

Usage:
    python create_admin.py --gitlab_username=amruthjakku --name="Amruth Jakku" --email="amruth@example.com"

This script creates the first admin user in the database.
"""

import argparse
import sys
import os
from pymongo import MongoClient
from datetime import datetime
import urllib.parse

def get_mongodb_uri():
    """Get MongoDB URI from environment or use default"""
    # Try to read from .env.local file
    env_file = os.path.join(os.path.dirname(__file__), '..', '.env.local')
    if os.path.exists(env_file):
        with open(env_file, 'r') as f:
            for line in f:
                if line.startswith('MONGODB_URI='):
                    return line.split('=', 1)[1].strip()
    
    # Fallback to environment variable
    return os.getenv('MONGODB_URI', 'mongodb://localhost:27017/internship_tracker')

def create_admin(gitlab_username, name, email, gitlab_id=None):
    """Create the first admin user"""
    
    # Connect to MongoDB
    mongodb_uri = get_mongodb_uri()
    client = MongoClient(mongodb_uri)
    
    try:
        # Get database name from URI or use default
        if 'mongodb+srv://' in mongodb_uri or 'mongodb://' in mongodb_uri:
            # Extract database name from URI
            parsed = urllib.parse.urlparse(mongodb_uri)
            db_name = parsed.path.lstrip('/') if parsed.path else 'internship_tracker'
        else:
            db_name = 'internship_tracker'
        
        db = client[db_name]
        users_collection = db.users
        
        # Check if admin already exists
        existing_admin = users_collection.find_one({
            'gitlabUsername': gitlab_username.lower()
        })
        
        if existing_admin:
            print(f"❌ Admin with GitLab username '{gitlab_username}' already exists!")
            return False
        
        # Check if any admin exists
        existing_admins = users_collection.count_documents({'role': 'admin'})
        
        # Create admin user
        admin_user = {
            'gitlabUsername': gitlab_username.lower(),
            'gitlabId': gitlab_id or f"admin_{gitlab_username.lower()}",
            'name': name,
            'email': email.lower(),
            'role': 'admin',
            'assignedBy': 'system_bootstrap',
            'isActive': True,
            'profileImage': None,
            'lastLoginAt': None,
            'createdAt': datetime.utcnow(),
            'updatedAt': datetime.utcnow()
        }
        
        # Insert admin user
        result = users_collection.insert_one(admin_user)
        
        if result.inserted_id:
            print(f"✅ Successfully created admin user!")
            print(f"   GitLab Username: {gitlab_username}")
            print(f"   Name: {name}")
            print(f"   Email: {email}")
            print(f"   Role: admin")
            print(f"   User ID: {result.inserted_id}")
            
            if existing_admins == 0:
                print(f"🎉 This is the first admin user in the system!")
            else:
                print(f"📊 Total admins in system: {existing_admins + 1}")
            
            return True
        else:
            print(f"❌ Failed to create admin user!")
            return False
            
    except Exception as e:
        print(f"❌ Error creating admin: {str(e)}")
        return False
    finally:
        client.close()

def main():
    parser = argparse.ArgumentParser(
        description='Create the first admin user for Internship Tracker'
    )
    parser.add_argument(
        '--gitlab_username',
        required=True,
        help='GitLab username of the admin user'
    )
    parser.add_argument(
        '--name',
        required=True,
        help='Full name of the admin user'
    )
    parser.add_argument(
        '--email',
        required=True,
        help='Email address of the admin user'
    )
    parser.add_argument(
        '--gitlab_id',
        help='GitLab user ID (optional, will be auto-generated if not provided)'
    )
    
    args = parser.parse_args()
    
    print("🚀 Creating admin user for Internship Tracker...")
    print(f"   GitLab Username: {args.gitlab_username}")
    print(f"   Name: {args.name}")
    print(f"   Email: {args.email}")
    print()
    
    success = create_admin(
        gitlab_username=args.gitlab_username,
        name=args.name,
        email=args.email,
        gitlab_id=args.gitlab_id
    )
    
    if success:
        print()
        print("🎯 Next Steps:")
        print("1. Start your Next.js application")
        print("2. Login with your GitLab account")
        print("3. You'll be automatically recognized as an admin")
        print("4. Use the admin dashboard to add mentors and colleges")
        sys.exit(0)
    else:
        print()
        print("❌ Failed to create admin user. Please check the error above.")
        sys.exit(1)

if __name__ == '__main__':
    main()
